easybuild/easyblocks/c/clang_aomp.py

```diff
@@ -0,0 +1,239 @@
+        # Mapping from known ROCm components to their configure method
+        self.cfg_method = {
+            'llvm-project': self._configure_llvm,
+            'ROCm-Device-Libs': self._configure_device_libs,
+            'Clang-OpenMP': self._configure_omp,
+            'aomp-extras': self._configure_aomp_extras,
+        }
+        # Prepare configuration options that point to the expected Clang build
+        self.llvm_compiler_flags = [
+            "-DCMAKE_C_COMPILER=%s" % os.path.join(self.installdir, 'bin', 'clang'),
+            "-DCMAKE_CXX_COMPILER=%s" % os.path.join(self.installdir, 'bin', 'clang++'),
+            "-DLLVM_INSTALL_PREFIX=%s" % self.installdir,
+            "-DLLVM_DIR=%s" % self.installdir,
+        ]
+        # Variables to be filled in the prepare step
+        self.amd_gfx_archs = []
+        self.cuda_archs = []
+        self.device_lib_path = None
+
+    def prepare_step(self, *args, **kwargs):
+        """
+        Setup target architectures for LLVM
+        """
+        super(EB_Clang_AOMP, self).prepare_step(*args, **kwargs)
+        # Detect CPU architecture and setup build targets for LLVM
+        cpu_arch = get_cpu_architecture()
+        if cpu_arch not in LLVM_ARCH_MAP:
+            raise EasyBuildError('Unknown CPU architecture for LLVM: %s' % cpu_arch)
+        self.target_archs.append(LLVM_ARCH_MAP[cpu_arch][0])
+        # If CUDA is loaded when building, try to build CUDA offload backend
+        if get_software_root('CUDA'):
+            self.target_archs.append('NVPTX')
+            ec_cuda_cc = self.cfg['cuda_compute_capabilities']
+            cfg_cuda_cc = build_option('cuda_compute_capabilities')
+            cuda_cc = cfg_cuda_cc or ec_cuda_cc or []
+            if not cuda_cc:
+                raise EasyBuildError("Can't build Clang-AOMP with CUDA support "
+                                     "without specifying 'cuda-compute-capabilities'")
+            self.cuda_archs = [cc.replace('.', '') for cc in cuda_cc]
+            self.log.info("Building offload support for the following CUDA architectures: '%s'",
+                          ' '.join(self.amd_gfx_archs))
+        self.log.info("Building LLVM for the following architectures: '%s'", ';'.join(self.target_archs))
+        # Setup AMD GFX list to build for
+        if self.cfg['gfx_list']:
+            self.amd_gfx_archs = self.cfg['gfx_list']
+        else:
+            self.amd_gfx_archs = DEFAULT_GFX_ARCHS
+        self.log.info("Building offload support for the following AMD architectures: '%s'",
+                      ' '.join(self.amd_gfx_archs))
+        # Ensure necessary libraries are downloaded and can be found
+        device_lib_dir_pattern = os.path.join(self.builddir, 'ROCm-Device-Libs-*')
+        hits = glob.glob(device_lib_dir_pattern)
+        if len(hits) == 1:
+            self.device_lib_path = hits[0]
+        else:
+            raise EasyBuildError("Could not find 'ROCm-Device-Libs' source directory")
```

Comment:
@nordmoen It's better to move this to `configure_step`, otherwise `--module-only` will always fail hard (since sources are not unpacked then, but `prepare_step` is run).

Reply:
Done =]